arch: arm: rom_start relocation configuration

jlh-makeen opened the pull request:
In order to support Linux rproc loading, some SOCs require the boot-vector and irq-vectors to be placed into a defined memory area for the mcu to boot.

This is necessary for NXP's IMX SOCs for instance but can be leveraged by other SOCs that have multiple
zephyr,flash choices.

This PR answers the discussion #71617

JiafeiPan commented on arch/arm/Kconfig:
```diff
@@ -34,6 +34,67 @@ config ARM_CUSTOM_INTERRUPT_CONTROLLER
+		   IMX8MN the boot-vector is ITCM (0x007E0000, aliased at 0x0).
+
+		  Example of DT definition:
+		  	$(dt_nodelabel_reg_addr_hex,ocram_s_sys)
```
Please fix compliance issue firstly:
arch/arm/Kconfig:81: space before tab in indent.

jlh-makeen replied:
Thanks for the review, I will :)

Is there a way to run compliance checks locally before submitting the PR ?
Because I wasn't warned by ./scripts/checkpatch.pl at the time of submission.

jlh-makeen replied:
I think I found the way to check the errors:

```
cd ~/zephyrproject/zephyr
./script/ci/check_compliance.py 
```

jlh-makeen replied:
> I think I found the way to check the errors:
> 
> ```
> cd ~/zephyrproject/zephyr
> ./script/ci/check_compliance.py 
> ```

And pushed the corrections :)

JiafeiPan commented:
@danieldegrasse @dlech please help to review this PR, thanks.

JiafeiPan commented on dts/arm/nxp/nxp_imx6sx_m4.dtsi:
```diff
@@ -29,9 +29,10 @@
+	tcml:memory@0 {
 		compatible = "nxp,imx-itcm";
-		reg = <0x1fff8000 DT_SIZE_K(32)>;
+		reg = <0x00000000 DT_SIZE_K(32)>;
```
Let's done change it to zero address, with startrom relocation, the size of zephyr.bin will be 512M bytes.

JiafeiPan replied:
@jlh-makeen , revert this update?

JiafeiPan commented on soc/nxp/imx/imx7d/Kconfig.defconfig.mcimx7d_m4:
```diff
@@ -11,4 +11,11 @@ config SYS_CLOCK_HW_CYCLES_PER_SEC
+#to translate it to OCRAM_S address
+#Select the relocation if the chosen flash is not the OCRAM_S
+config ROMSTART_RELOCATION_ROM
+	default y if $(dt_chosen_reg_addr_hex,$(DT_CHOSEN_Z_FLASH))!=$(dt_nodelabel_reg_addr_hex,ocram_s_code)
```
As the size of zephyr.bin may be enlarged, and romstart relocation is not necessary for some usecase even in case of startrom is not located at ocram_s,  so let's remove this enablement by dts checking, and enable CONFIG_ROMSTART_RELOCATION_ROM by manual, that is to say application decide whether enable it or not, so by default it is disabled.

But we can set default remstart relocation address just like:
config ROMSTART_REGION_ADDRESS
           default $(dt_nodelabel_reg_addr_hex,ocram_s_code)

jlh-makeen replied:
Yes the size of bin is enlarged, but the bin file can only be loaded with imx_fwloader or u-boot which are loaders that explicitly copy the boot-vector into OCRAM_S (imx7d).

I can remove the automatic enabling here if it's creating trouble. 

I will split the commits :)

jlh-makeen replied:
@JiafeiPan  I don't know why CI failed on the last push as local compliance check and checkpatch were okay.
Just rebased on main and pushed again to retrig the CI checks.

JiafeiPan reviewed (changes requested):
And please split the update in multiple commits,  the common code is one commit, and the update for platform is another one.

jlh-makeen committed:
arch: arm: rom_start relocation configuration

In order to support Linux rproc loading, some SOCs require
the boot-vector and irq-vectors to be placed into a defined
memory area for the mcu to boot.

This is necessary for NXP's IMX SOCs for instance but
can be leveraged by other SOCs that have multiple
zephyr,flash choices.

Signed-off-by: Jérémy LOCHE - MAKEEN Energy <jlh@makeenenergy.com>

jlh-makeen committed:
nxp: imx7d-6sx: add rom_start relocation

Add the Kconfig options and use the aliased
addresses for the bootcode regions of the IMX7D
and IMX6SX SOCs to allow the Linux rproc
framework to load the irq-vectors into
the correct memory areas.

Activating this option might enlarge the bin
file if the zephyr,flash and rom_start chosen
region addresses are not matching.

It is up to the user to enable this feature
based on code location choices (OCRAM, DDR, TCM...).

Signed-off-by: Jérémy LOCHE - MAKEEN Energy <jlh@makeenenergy.com>

JiafeiPan commented on dts/arm/nxp/nxp_imx7d_m4.dtsi:
```diff
@@ -56,9 +56,10 @@
+		ocram_s_code: code@0 {
 			compatible = "nxp,imx-code-bus";
-			reg = <0x20180000 DT_SIZE_K(32)>;
+			reg = <0x00000000 DT_SIZE_K(32)>;
```
@jlh-makeen , revert this update?

jlh-makeen replied:
I can explain this choice:

The Linux imx_rproc driver is NOT mapping the 0x20180000 address space to 0x00180000, which means that when user selects zephyr,flash=&ocram_s_code, rproc doesn't load the irq_vector and code at the right memory space. 
Even though the M4 itself understands 0x20180000 which is the CODE bus where instructions are placed, rproc will not load the elf-file correctly. 
This is because OCRAM_S is at address 0x00180000 on the A7 side.

By using address 0, we are guaranteed that all CODE addresses/pointers will be aliased to the OCRAM_S_CODE address by the M4 and to 0x00180000 by the imx_rproc driver.

That reasoning is valid for imx7d, imx6sx and other soc like imx8

https://github.com/nxp-imx/linux-imx/blob/lf-6.6.y/drivers/remoteproc/imx_rproc.c#L262

Unless I am proven wrong and explained why using the aliased address instead of the raw address here is not a good idea, I'd rather keep this changeset :)

JiafeiPan replied:
Yes, indeed, actually we also need to add 0x20180000 mapping in imx_rproc driver as well, but currently both 0x00000000 and 0x00180000 are okay at here.

JiafeiPan commented on arch/arm/Kconfig:
```diff
@@ -34,6 +34,67 @@ config ARM_CUSTOM_INTERRUPT_CONTROLLER
+
+	  In general this option should be chosen if the zephyr,flash chosen node
+	  is not placed into the boot-vector memory area
+
```
Add the description of impaction? for example: If this item is enabled, the zephyr.bin image maybe enlarged, if only zephyr.bin is used as booting image, no need to enable it.

jlh-makeen replied:
@JiafeiPan Yes this is a relevant place to put it indeed, I'll provide a small update of this doc tomorrow 👍

jlh-makeen replied:
@JiafeiPan Commit is now reworked with explainations :)

JiafeiPan reviewed (approved):
looks fine for me.

jlh-makeen commented:
@ithinuel @dleach02 @mmahadevan108 Anyone willing to review this PR :-)

ithinuel reviewed (approved):
No objections from me.
